x/logic/fs/composite/fs_test.go

```diff
@@ -0,0 +1,243 @@
+			{
+				files:     map[string][]fileSpec{"vfs2": {vfs2File1}},
+				file:      "vfs2:///file1",
+				fsType:    fsTypeReadFile,
+				wantError: "vfs2:///file1: file is corrupted",
+				want:      nil,
+			},
+			{
+				files:     map[string][]fileSpec{"vfs2": {vfs2File1}},
+				file:      "vfs2:///file1",
+				fsType:    fsTypeFile,
+				wantError: "vfs2:///file1: file is corrupted",
+				want:      nil,
+			},
+		}
+
+		for nc, tc := range cases {
+			Convey(fmt.Sprintf("Given the test case #%d - file %s", nc, tc.file), func() {
+				Convey("and a bunch of mocked file systems", func() {
+					vfss := make(map[string]fs.FS)
+					for k, files := range tc.files {
+						switch tc.fsType {
+						case fsTypeFile:
+							vfs := testutil.NewMockFS(ctrl)
+							for _, f := range files {
+								registerFileToFS(vfs, f.name, f.content, f.modTime, f.isCorrupted)
+							}
+							vfss[k] = vfs
+						case fsTypeReadFile:
+							vfs := testutil.NewMockReadFileFS(ctrl)
+							for _, f := range files {
+								registerFileToReadFileFS(vfs, f.name, f.content, f.modTime, f.isCorrupted)
+							}
+							vfss[k] = vfs
+						default:
+							panic("unsupported fs type")
+						}
+					}
+
+					Convey("and a composite file system under test", func() {
+						compositeFS := NewFS()
+						for mountPoint, vfs := range vfss {
+							compositeFS.Mount(mountPoint, vfs)
+						}
+
+						Convey(fmt.Sprintf(`when the open("%s") is called`, tc.file), func() {
+							result, err := compositeFS.Open(tc.file)
+
+							Convey("then the result should be as expected", func() {
+								if tc.wantError == "" {
+									So(err, ShouldBeNil)
+
+									stat, _ := result.Stat()
+									So(stat.Name(), ShouldEqual, tc.file)
+									So(stat.Size(), ShouldEqual, int64(len(tc.want.content)))
+									So(stat.ModTime(), ShouldEqual, tc.want.modTime)
+									So(stat.IsDir(), ShouldBeFalse)
+								} else {
+									So(err, ShouldNotBeNil)
+									So(err.Error(), ShouldEqual, fmt.Sprintf("open %s", tc.wantError))
+								}
+							})
+						})
+
+						Convey(fmt.Sprintf(`when the readFile("%s") is called`, tc.file), func() {
+							result, err := compositeFS.ReadFile(tc.file)
+
+							Convey("Then the result should be as expected", func() {
+								if tc.wantError == "" {
+									So(err, ShouldBeNil)
+									So(result, ShouldResemble, tc.want.content)
+								} else {
+									So(err, ShouldNotBeNil)
+									So(err.Error(), ShouldEqual, fmt.Sprintf("readfile %s", tc.wantError))
+								}
+							})
+						})
+
+						Convey(`when the ListMounts() is called`, func() {
+							result := compositeFS.ListMounts()
+
+							Convey("Then the result should be as expected", func() {
+								vfssNames := make([]string, 0, len(vfss))
+								for k := range vfss {
+									vfssNames = append(vfssNames, k)
+								}
+								sort.Strings(vfssNames)
+								So(result, ShouldResemble, vfssNames)
+							})
+						})
+					})
+				})
+			})
+		}
+	})
+}
```

Comment:
The table driven approach is already used here. Sure, the length of the function is consequential, but because of the [goconvey](https://github.com/smartystreets/goconvey) style of testing, the tests are segmented and self-describing (as could be obtained by breaking them down by function).